src/lib.rs

```diff
@@ -351,3 +435,71 @@ fn convert_response_to_query_response(res: skar_client::QueryResponse) -> Result
         },
     })
 }
+
+fn convert_response_to_pyarrow<'py>(
+    py: Python<'py>,
+    res: skar_client::QueryResponse,
+) -> Result<QueryResponseArrow> {
+
+    let blocks = res.data.blocks;
+    let transactions = res.data.transactions;
+    let logs = res.data.logs;
+
+    let (blocks, transactions, logs) = Python::with_gil(|py| {
+        let pyarrow = py.import("pyarrow.ipc")?;
+        let blocks = convert_batch_to_pyarrow(py, pyarrow, blocks)?;
+        let transactions = convert_batch_to_pyarrow(py, pyarrow, transactions)?;
+        let logs = convert_batch_to_pyarrow(py, pyarrow, logs)?;
+        Ok::<(Vec<PyObject>, Vec<PyObject>, Vec<PyObject>), PyErr>((blocks, transactions, logs))
+    })?;
+
+    Ok(QueryResponseArrow {
+        archive_height: res
+            .archive_height
+            .map(|h| h.try_into())
+            .transpose()
+            .context("convert height")?,
+        next_block: res.next_block.try_into().context("convert next_block")?,
+        total_execution_time: res
+            .total_execution_time
+            .try_into()
+            .context("convert total_execution_time")?,
+        data: QueryResponseDataArrow {
+            blocks,
+            transactions,
+            logs,
+        },
+    })
+}
+
+fn convert_batch_to_pyarrow<'py>(
+    py: Python<'py>,
+    pyarrow: &'py PyModule,
+    batches: Vec<ArrowBatch>,
+) -> PyResult<Vec<PyObject>> {
+    let mut list = Vec::new();
+    let schema = batches.first().unwrap().schema.fields.clone();
+    assert_eq!(batches.first().unwrap().chunk.arrays().len(), schema.len());
+
+    // TODO: probably could avoid to_boxed usage and consume array somehow to avoid allocation
+    let data = vec![vec![]; batches.first().unwrap().chunk.arrays().len()];
+    for batch in batches {
+        for (index, array) in batch.chunk.arrays().iter().enumerate() {
+            data[index].push(array.to_boxed());
+        }
+    }
+
+    for (array, field) in data.into_iter().zip(schema.into_iter()) {
+        let field = Field::new(field.name.clone(), DataType::Struct(vec![field]), true);
```

Comment:
This should be essentially `DataType::Struct(schema.fields.clone(), true)` to make the types right